### DISCUSSION TITLE
How to interrupt/stop the main thread and have to do something else

### DISCUSSION THREAD

**Discussion by GM-Script-Writer-62850** [micropython/micropython#12026] [General]
Issues i am trying to work around: 

* PICO W's wifi is only accessible on core 0 in the main thread (can't be nigher be accessed on core 1 nor on core 0 via a interrupt)
* `uasyncio` can only be used on a single core as far as i know so if it is on core 1 it can't be used for a server on a pico w
* urequest is a blocking function so I do not want to use in on my `uasyncio` core
* [uaiohttpclient.py](https://github.com/micropython/micropython-lib/blob/master/micropython/uaiohttpclient/uaiohttpclient.py) does not support post data
* The W5500-EVB-PICO can use the network on both cores, but this has been a stability nightmare when using _thread and DHT22s

Lets assume this example is running on core 0 with `uasyncio` on core 1:
[3.9.1. A simple server for static pages](https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf#%5B%7B%22num%22%3A24%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C115%2C841.89%2Cnull%5D) (not planning on doing anything heavy, just need to receive a few bytes of data)

Most of the time will be spent on line 54 waiting on a client, i can trigger a interrupt from core 1 by toggling output pin tied to a irq, but said interrupt can't use the NIC and [_thread.interrupt_main()](https://docs.python.org/3/library/_thread.html) is not implemented

With `s.accept()` in progress i can't just have a loop and check if there is something else to do, any idea of how to work around the limitations? the only thing i can think of would be upgrading `uaiohttpclient.py` thus not needing urequest to run on core 0 allowing my to use `uasyncio` on core 0 where it can run a sever and use `uaiohttpclient.py`  then i can just use core 1 for slow stuff like reading sensors like DHT22s

**Comment by peterhinch** [comment]
A possible approach is to run `uasyncio` on core 0 and delegate blocking functions to core 1, as described [here](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md#4-taming-blocking-functions).

**Reply by GM-Script-Writer-62850** [reply]
in the case of blocking functions like urequest when using a pico w this MUST be on core 0 in the main thread or it will fail

**Comment by karfas** [comment]
> urequest is a blocking function so I do not want to use in on my uasyncio core
> [uaiohttpclient.py](https://github.com/micropython/micropython-lib/blob/master/micropython/uaiohttpclient/uaiohttpclient.py) does not support post data

In my opinion, your usage of (blocking) urequest is the main problem here. 
Why not hack uaiohttpclient to support POST requests ? At least for small amounts of data (where you can calculate the `Content-Length:` header) POST is not so much different from GET.

**Reply by GM-Script-Writer-62850** [reply]
https://pastebin.com/fWPJJrVh

Not sure about performance of lines 60 and 61 is there a better way

edit and now i need to patch it for use of not port 80, i could just hard code it to port 8080...

**Comment by rkompass** [comment]
In the reference you gave there is also [3.9.3. An asynchronous web server](https://datasheets.raspberrypi.com/picow/connecting-to-the-internet-with-pico-w.pdf). Did you try that too? And isn't it what you need?
Or can you explain, why not?

**Reply by GM-Script-Writer-62850** [reply]
> * urequest is a blocking function so I do not want to use in on my uasyncio core
> * PICO W's wifi is only accessible on core 0 in the main thread
> * [uaiohttpclient.py](https://github.com/micropython/micropython-lib/blob/master/micropython/uaiohttpclient/uaiohttpclient.py) does not support post data

i'd love to use 2 instances of uasyncio, but that is not supported

**Comment by Carglglz** [comment]
@GM-Script-Writer-62850 I made an [async_urequest.py](https://github.com/Carglglz/asyncmd/blob/main/async_modules/async_urequests/async_urequests.py) version that may be of help (just delete/comment the `ssl` keyword arg [here](https://github.com/Carglglz/asyncmd/blob/74ec11167ae6477ba924b5c020522e3cfb82fd74/async_modules/async_urequests/async_urequests.py#L115)). I haven't tested the POST method but I guess it should work 👍🏼

**Reply by GM-Script-Writer-62850** [reply]
meanwhile i was patching... [uaiohttpclient.py](https://pastebin.com/fWPJJrVh) did not test it yet
